plugins/identity-shiro/src/main/java/org/opensearch/identity/shiro/ShiroSubject.java

```diff
@@ -85,7 +85,13 @@ public String toString() {
      */
     public void authenticate(AuthToken authenticationToken) {
         final org.apache.shiro.authc.AuthenticationToken authToken = authTokenHandler.translateAuthToken(authenticationToken)
-            .orElseThrow(() -> new UnsupportedAuthenticationToken());
-        shiroSubject.login(authToken);
+            .orElseThrow(UnsupportedAuthenticationToken::new);
+        wrappedSubject.login(authToken);
+    }
+
+    @SuppressWarnings("unchecked")
+    @Override
+    public Optional<Principal> getApplication() {
```

Comment:
> This is included because it is an optional defined in the Subject interface. 
I mean this is should be designed as exclusive abstractions, the optional just branch off the different states and makes API prone to misuse.

Reply:
On a Subject there the notion of a principal.  In different frameworks it is represented as, an ordered collections of principals, other have multiple principals with one called a primary, or in what is proposed here there is a user principle and a separate application principal.  What do you think?

I like the approach to have a distinct application principal as OpenSearch application concept for plugin and extension systems is limited to core.  The Application manager that @scrawfor99 has proposed isn't meant to be accessed/extended by extensions.

Reply:
> other have multiple principals with one called a primary, or in what is proposed here there is a user principle and a separate application principal. What do you think?

I don't see that in the current design: every single (meaningful) usage of the `getApplication()` delegates to `getPrincipal()` (see https://github.com/opensearch-project/OpenSearch/pull/7850/files#diff-1aa331c528b9d022fdd78cc985d90857b10cf9d0d5da28bafd2acb5dbbf799e3R166 fe),  there is no primary or secondary. This is why I think we have to split those into distinct subjects: user, application (may be later plugin or extension).

Reply:
I've revisited many of these usages and largely removed subject's usage.  There is another pull request [[link]](https://github.com/opensearch-project/OpenSearch/pull/8526) where we are adding service accounts which are tied to applications, and allow them to make REST requests - IMO this is a better personification as it allows applications to  follow the same external authentication processes as users.

Contrived example of how the subject is changable over time;
So if I make a request with my openSearch client, after its authenticated, its `{sub: peternied}`, and then it flows into a plugins action handler, it should be `{sub: peternied, app: machine_learning}`, then after the request leaves the handler, it becomes just `{sub: peternied}`, if it enters another plugin entry point `{sub: peternied, app: security_plugin}` gets changed.  The primary subject is still me; however, the application changes as the request is handled in OpenSearch.

peternied, machine_learning, security_plugin are all principals, but you cannot authenticate with the cluster as machine_learning so it isn't a Subject.  Let me know if this helps clarify

Reply:
Oh I see, thanks @peternied , it cleared the confusion for me, so this is basically an originator, right? Should we call it as such?

Reply:
Good pull on originator [1] is the correct name.  I think on the subject, we would switch to getOriginator() as you suggest, but keep the ApplicationAwareSubject which handle when the originator is an plugin/extension.

- [1] https://csrc.nist.gov/glossary/term/originator